File: problems/0327-count-of-range-sum/solution.cpp
// Link: https://leetcode.cn/problems/count-of-range-sum/
// Level: Hard
// Date: 2025-04-19

class FenwickTree {
    private:
        vector<int> tree;
        int n;
    public:
        FenwickTree(int size) : n(size), tree(size + 1) {}

        void update(int index, int delta) {
            while (index <= n) {
                tree[index] += delta;
                index += index & -index;
            }
        }

        int query(int index) {
            int res = 0;
            while (index > 0) {
                res += tree[index];
                index -= index & -index;
            }
            return res;
        }
};

class Solution {
    public:
        int countRangeSum(vector<int>& nums, int lower, int upper) {
            long long sum = 0;
            vector<long long> preSum = {0};
            for (int num : nums) {
                sum += num;
                preSum.push_back(sum);
            }

            vector<long long> allNumbers;
            for (long long x : preSum) {
                allNumbers.push_back(x);
                allNumbers.push_back(x - lower);
                allNumbers.push_back(x - upper);
            }
            sort(allNumbers.begin(), allNumbers.end());
            auto last = unique(allNumbers.begin(), allNumbers.end());
            allNumbers.erase(last, allNumbers.end());

            FenwickTree ft(allNumbers.size());
            int res = 0;
            for (long long s_val : preSum) {
                int left_idx = lower_bound(allNumbers.begin(), allNumbers.end(), s_val - upper) - allNumbers.begin() + 1;
                int right_idx = lower_bound(allNumbers.begin(), allNumbers.end(), s_val - lower) - allNumbers.begin() + 1;

                if (right_idx > 0 && left_idx <= right_idx) {
                    res += ft.query(right_idx) - (left_idx > 1 ? ft.query(left_idx - 1) : 0);
                }
                int s_pos = lower_bound(allNumbers.begin(), allNumbers.end(), s_val) - allNumbers.begin() + 1;
                ft.update(s_pos, 1);
            }
            return res;
        }
};

File: problems/0327-count-of-range-sum/solution.py
# Link: https://leetcode.cn/problems/count-of-range-sum/
# Level: Hard
# Date: 2025-04-19

class FenwickTree:
    def __init__(self, size):
        self.n = size
        self.tree = [0] * (size + 1)

    def update(self, index, delta):
        while index <= self.n:
            self.tree[index] += delta
            index += index & -index

    def query(self, index):
        result = 0
        while index > 0:
            result += self.tree[index]
            index -= index & -index
        return result

class Solution:
    def countRangeSum(self, nums: List[int], lower: int, upper: int) -> int:
        pre_sum = [0]
        s = 0
        for num in nums:
            s += num
            pre_sum.append(s)

        all_numbers = set()
        for s_val in pre_sum:
            all_numbers.add(s_val)
            all_numbers.add(s_val - lower)
            all_numbers.add(s_val - upper)

        sorted_all = sorted(all_numbers)
        value_to_idx = {v : i + 1 for i, v in enumerate(sorted_all)}

        ft = FenwickTree(len(sorted_all))
        res = 0
        for s_val in pre_sum:
            cur_left = s_val - upper
            cur_right = s_val - lower

            left_idx = value_to_idx[cur_left]
            right_idx = value_to_idx[cur_right]

            res += ft.query(right_idx) - ft.query(left_idx - 1)
            s_pos = value_to_idx[s_val]
            ft.update(s_pos, 1)

        return res
